### PR TITLE
add Einsums

### PR DESCRIPTION
**Pull request by loriab**
Einsums is a C++ library that does tensor operations and determines at compile time the best blas pattern to call. It is being used in some new computational chemistry development.

A possible concern is Einsums's handling of two dependencies: h5cpp (header-only) and hptt (compiled; 11 files). Ordinarily, I'd package these up with new recipes. In this case, I haven't because of the following:
* upstream hptt (https://github.com/springer13/hptt) hasn't been worked on since 2018 and isn't merging fixes. upstream h5cpp (https://github.com/steven-varga/h5cpp) hasn't been worked on since 2021 and isn't merging fixes.
* h5cpp and hptt aren't packaged by Fedora, Debian, or OpenSuse.
* Einsums needs modifications to both dependencies, both for build and content.

Instead, Einsums:
* maintains forks of the upstream dependencies (https://github.com/Einsums) and pulls in the src at build-time, checking the hash.
* h5cpp (the header-only) is consumed at Einsums compile-time,  but in the Einsums install, its headers are installed under the Einsums namespace for downstream packages to compile against Einsums.
* hptt gets compiled as an object library into Einsums and doesn't appear in an Einsums install.
* the licenses for both h5cpp and hptt are included in the THIRD_PARTY_LICENSES file.

I hope these reasons and practices are sufficient to allow vendoring these particular dependencies in Einsums.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
  * see notes above. yes, licenses are packaged.
- [x] If static libraries are linked in, the license of the static library is packaged.
  * see notes above. yes, license is packaged. 
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by jturney** [comment]
I'm willing to be a maintainer for this package.

**Comment by loriab** [comment]
@conda-forge/help-c-cpp, ready for review!

File: recipes/einsums/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  host:
+    - catch2
+    - elfutils                                                             # [linux]
+    - fmt 10.1
+    - hdf5
+    - range-v3 0.12
+    - zlib
+    - mkl-devel {{ mkl }}                                                  # [blas_impl == "mkl"]
+    - blas-devel * *_{{ blas_impl }}                                       # [blas_impl == "openblas"]
+    - fftw                                                                 # [blas_impl == "openblas"]
```
Comment (jakirkham) [inline recipes/einsums/meta.yaml:41]: Is this meant to be constrained by `openblas`?

**Review by jakirkham — COMMENTED** [review]
Thanks Lori! 🙏

Had a question below

File: recipes/einsums/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    - openblas * openmp_*                                                  # [blas_impl == "openblas"]
+    - llvm-openmp
```
Comment (jakirkham) [inline recipes/einsums/meta.yaml:43]: As [`llvm-openmp` has `run_exports/strong`]( https://github.com/conda-forge/openmp-feedstock/blob/62f174e38a237b9be6fd86d7553ae55fbbee6505/recipe/meta.yaml#L30-L32 ), there is no need to list it in `host`

```suggestion
```

File: recipes/einsums/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  run:
+    - libopenblas * openmp_*                                               # [blas_impl == "openblas"]
+    - llvm-openmp
```
Comment (jakirkham) [inline recipes/einsums/meta.yaml:46]: Same as above with `run`

```suggestion
```

File: recipes/einsums/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    - elfutils                                                             # [linux]
+    - libzlib
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - ninja
+    - llvm-openmp
```
Comment (jakirkham) [inline recipes/einsums/meta.yaml:31]: As LLVM OpenMP is used with macOS builds and GNU OpenMP is used on Linux, would change these like so

```suggestion
    - llvm-openmp                                                          # [osx]
    - libgomp                                                              # [linux]
```

**Review by jakirkham — COMMENTED** [review]
Sorry for the second review. Noticed some things with OpenMP usage and added them below

File: recipes/einsums/meta.yaml (hunk 1 above)
Comment (loriab) [reply recipes/einsums/meta.yaml:41]: Are you asking about fftw in particular being constrained by openblas? In that case, yes. einsums uses the internal FFT when MKL is selected, and MKL & OpenBLAS are the only implementations allowed at present.

Thanks for the review! Sorry I wasn't monitoring this PR last week. I'll fix up the OpenMP items you raised.

**Comment by loriab** [comment]
Ok, the suggestions didn't work as proposed for mkl:
```
Packaging einsums-0.2-mkl_h09027b7_0
...
  ERROR (einsums,lib/libeinsums.so.0.2.0): Needed DSO lib/libomp.so found in ['llvm-openmp']
  ERROR (einsums,lib/libeinsums.so.0.2.0): .. but ['llvm-openmp'] not in reqs/run, (i.e. it is overlinking) (likely) or a missing dependency (less likely)
WARNING (einsums): run-exports library package conda-forge::_openmp_mutex-4.5-2_kmp_llvm in requirements/run but it is not used (i.e. it is overdepending or perhaps statically linked? If that is what you want then add it to `build/ignore_run_exports`)
```
probably b/c `mkl` has depends `_openmp_mutex * *_llvm, _openmp_mutex >=4.5, llvm-openmp >=16.0.6` but the analyzer wants openmp to be explicit. So I added back `llvm-openmp` only for mkl+linux.

The osx issue is the new `libxcc` that's popped up a lot lately.

File: recipes/einsums/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    - zlib
+    - mkl-devel                                                            # [blas_impl == "mkl"]
+    - blas-devel * *_openblas                                              # [blas_impl == "openblas"]
+  files:
+    - timing/CMakeLists.txt
+  source_files:
+    - timing/main.cpp
+  commands:
+    # Verify library
+    - test -f $PREFIX/lib/libeinsums$SHLIB_EXT                             # [unix]
```
Comment (carterbox) [inline recipes/einsums/meta.yaml:64]:
```suggestion
    # Verify library
    - test -f $PREFIX/lib/libeinsums.so.{{ version }}                      # [linux]
    - test -f $PREFIX/lib/libeinsums$SHLIB_EXT                             # [unix]
```

Comment (carterbox) [reply recipes/einsums/meta.yaml:64]: My only suggestion is to start checking the explicit so version to prevent ABI breaks in the future.

Comment (loriab) [reply recipes/einsums/meta.yaml:64]: I'll add in the SO check. Thanks for the merge!

**Pull request merged**